programs/locked-voter/src/instructions/open_partial_unstaking.rs

```diff
@@ -0,0 +1,84 @@
+use crate::*;
+
+/// Accounts for [voter::open_partial_unstaking].
+#[derive(Accounts)]
+pub struct OpenPartialUnstaking<'info> {
+    /// [Locker].
+    #[account(mut)]
+    pub locker: Box<Account<'info, Locker>>,
+
+    /// [Escrow].
+    #[account(mut, has_one = locker, has_one = owner)]
+    pub escrow: Box<Account<'info, Escrow>>,
+
+    /// [Escrow].
+    #[account(
+        init,
+        payer = owner,
+        space = 8 + PartialUnstaking::LEN
+    )]
+    pub partial_unstake: Account<'info, PartialUnstaking>,
+
+    #[account(mut)]
+    pub owner: Signer<'info>,
+
+    /// System program.
+    pub system_program: Program<'info, System>,
+}
+
+impl<'info> OpenPartialUnstaking<'info> {
+    /// Creates a new [Locker].
+    pub fn open_partial_unstaking(&mut self, amount: u64) -> Result<()> {
+        let partial_unstake_pk = self.partial_unstake.key();
+        let escrow_pk = self.escrow.key();
+
+        require!(amount > 0, crate::ErrorCode::AmountIsZero);
+        require!(
+            amount <= self.escrow.amount, // todo check should be <= or <
+            crate::ErrorCode::InvalidAmountForPartialUnstaking
+        );
+
+        let partial_unstake = &mut self.partial_unstake;
+        let escrow = &mut self.escrow;
+
+        partial_unstake.escrow = escrow_pk;
+        partial_unstake.amount = amount;
+
+        let current_time = Clock::get()?.unix_timestamp;
+        let remaining_duration =
+            unwrap_int!(escrow.get_remaining_duration_until_expiration(current_time, &self.locker));
+        require!(remaining_duration > 0, crate::ErrorCode::EscrowHasBeenEnded); // no point for partial unstaking if escrow has been ended
```

Comment:
Actually, this is nearly impossible to be reached. The is reason because `get_remaining_duration_until_expiration` ` let duration = self.escrow_ends_at.checked_sub(current_time)?;` will overflow.

`saturating_sub` is better I think as it capped at 0.

Reply:
hmm, why ` this is nearly impossible to be reached`? if `escrow_ends_at < current_time` then remaining duration is zero, we also need to check for max_lock case too

Reply:
Sorry, my bad, missed `if self.escrow_ends_at < current_time {
            return Some(0);
        }`